src/Host/Broker/Impl/UserProfile/UserProfileManager.cs

```diff
@@ -0,0 +1,52 @@
+﻿using Microsoft.Common.Core;
+using Microsoft.Common.Core.Json;
+using Microsoft.Extensions.Logging;
+using Microsoft.R.Host.Protocol;
+using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.IO.Pipes;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.R.Host.Broker.UserProfile {
+    public class UserProfileManager {
+        private readonly ILogger _logger;
+
+        public UserProfileManager(ILogger<UserProfileManager> logger) {
+            _logger = logger;
+        }
+
+        public Task<RUserProfileServiceResponse> CreateProfileAsync(RUserProfileServiceRequest request, CancellationToken ct) {
+            return ProfileWorkerAsync(NamedPipeServerStreamFactory.CreatorName, Resources.Error_ProfileCreationFailedIO, request, ct);
+        }
+
+        public Task<RUserProfileServiceResponse> DeleteProfileAsync(RUserProfileServiceRequest request, CancellationToken ct) {
+            return ProfileWorkerAsync(NamedPipeServerStreamFactory.DeletorName, Resources.Error_ProfileDeletionFailedIO, request, ct);
+        }
+
+        private async Task<RUserProfileServiceResponse> ProfileWorkerAsync(string name, string log, RUserProfileServiceRequest request, CancellationToken ct) {
+            using (NamedPipeClientStream client = new NamedPipeClientStream(name)) {
+                try {
+                    await client.ConnectAsync(ct);
+
+                    string jsonReq = JsonConvert.SerializeObject(request);
```

Comment:
You can serialize directly to the stream if you wrap it into a `TextWriter`, via `JsonSerializer.Serialize` (and ditto for deserialization below with `TextWriter`, although you may need to add another helper to `Json` there). Then you don't need to deal with buffers, and it'll also properly handle the case when response is larger than 1K.